Scripts/Runtime/Collections/UnsafeBuffer.cs

```diff
@@ -0,0 +1,550 @@
+                blockPointer->Next = null;
+                
+                //Update lane writing info
+                m_Lane->WriterHead = blockPointer->Data;
+                m_Lane->WriterEndOfBlock = blockPointer->Data + m_BufferInfo->BlockSize;
+
+                //If this is the first block for the lane, we'll assign that
+                if (m_Lane->FirstBlock == null)
+                {
+                    m_Lane->CurrentWriterBlock = m_Lane->FirstBlock = blockPointer;
+                }
+                //Otherwise we'll update our linked list of blocks
+                else
+                {
+                    m_Lane->CurrentWriterBlock->Next = blockPointer;
+                    m_Lane->CurrentWriterBlock = blockPointer;
+                }
+            }
+        }
+        
+        //*************************************************************************************************************
+        // READERS
+        //*************************************************************************************************************
+        
+        /// <summary>
+        /// A lightweight struct to allow for reader from the collection
+        /// </summary>
+        [BurstCompatible]
+        public struct Reader
+        {
+            [NativeDisableUnsafePtrRestriction] private readonly BufferInfo* m_BufferInfo;
+            [NativeDisableUnsafePtrRestriction] private readonly Lane* m_Lanes;
+            
+            /// <summary>
+            /// How many lanes the collection has available to read from
+            /// </summary>
+            public int LaneCount
+            {
+                get => m_BufferInfo->LaneCount;
+            }
+            
+            internal Reader(ref UnsafeBuffer<T> unsafeBuffer)
+            {
+                m_Lanes = unsafeBuffer.m_Lanes;
+                m_BufferInfo = unsafeBuffer.m_BufferInfo;
+            }
+            
+            /// <summary>
+            /// Creates a <see cref="LaneReader"/> instance locked to a specific lane index.
+            /// </summary>
+            /// <param name="laneIndex">The lane index to use</param>
+            /// <returns>A <see cref="LaneReader"/> instance</returns>
+            public LaneReader CreateReaderForLane(int laneIndex)
+            {
+                Assert.IsTrue(laneIndex < m_BufferInfo->LaneCount && laneIndex >= 0);
+                
+                Lane* lane = m_Lanes + laneIndex;
+                return new LaneReader(lane, m_BufferInfo);
+            }
+        }
+        
+        /// <summary>
+        /// A lightweight reader instance locked to a specific lane.
+        /// <see cref="LaneReader"/>'s manage their own state so that multiple readers can read from a given lane at
+        /// the same time.
+        /// </summary>
+        [BurstCompatible]
+        public struct LaneReader
```

Comment:
Could maybe make this `readonly` if you captured the read head location in a struct and stored the pointer. Probably not worth it.

Reply:
I want to be able to have every instance of a `LaneReader` be able to read in parallel so we need to store state on the `LaneReader` itself.